### ISSUE TITLE
Problems with multiple pkcs11 readers

### ISSUE THREAD

**Issue by danepowell** [OpenSC/OpenSC#2741]
### Problem Description

I have two Yubikeys plugged into my machine at all times, each for a different organization. One implements pkcs11 and the other does not.

The problem is that opensc sees these keys as interchangeable, and during pkcs11 authentication will pick whichever was plugged in first (more or less).

I tried ignoring the non-pkcs11 key using the `ignored_readers` config option, but this won't work because the device names aren't even static. The first one plugged in is named `Yubico YubiKey OTP+FIDO+CCID`, the second is `Yubico YubiKey OTP+FIDO+CCID 01`. And since partial matches are accepted, even if the names were static I couldn't ignore the first one.

### Proposed Resolution

We need a way to either allow or ignore devices based on serial number, not just device name.

Seems similar to https://github.com/OpenSC/OpenSC/issues/397

**Comment by dengert** [comment]
> Yubico YubiKey OTP+FIDO+CCID, the second is Yubico YubiKey OTP+FIDO+CCID 01 are the names of the readers. 

What do you mean by: "One implements pkcs11 and the other does not." 

Are yo using PIV or OpenPGP on either or both Yubikeys?

We would need more information to tell these two devices apart. 

- pkcs11-tool --list-slots   # will list slot info about the two devices, and list the "manufacturer" which is a clue to which driver is being used. 
- pkcs11-tool --slot-index 0 -O   # list the "Slot 0 (0x0)" objects 
- pkcs11-tool --slot-index 1 -O   # list the "Slot 1 (0x4)" objects

If either both the tokens have both PIV and OpenPGP applets active, OpenSC can have a problem selecting PIV on one and OpenPGP on the other. There may be a way to select one on one and one on the other.

**Comment by danepowell** [comment]
> What do you mean by: "One implements pkcs11 and the other does not."

Sorry, let me explain. I've set up the "Foo PIV Certificate" (slot 0) token for pksc11 SSH auth (via `PKCS11Provider /usr/local/lib/opensc-pkcs11.so` in SSH config). The "PIV_II" (slot 1) token is unrelated and only used for browser-based auth.

The problem is there's no way to force OpenSC to use "Foo PIV Certificate" (or else ignore "PIV_II"). It seems to use whatever is in slot 0, and that can change depending on what order devices are registered.

```
$ pkcs11-tool --list-slots
Available slots:
Slot 0 (0x0): Yubico YubiKey OTP+FIDO+CCID
  token label        : Foo PIV Certificate
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : abcd1234
  pin min/max        : 4/8
Slot 1 (0x4): Yubico YubiKey OTP+FIDO+CCID 01
  token label        : PIV_II
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized, user PIN locked
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : 00000000
  pin min/max        : 4/8
```

**Comment by dengert** [comment]
Google for: ssh pkcs11 uri

RedHat has good explanation.

The URI could be used select the token label or serial number.

sorry no computer access right now to try it.

**Comment by danepowell** [comment]
I think you're referring to this page: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/configuring-applications-to-use-cryptographic-hardware-through-pkcs-11_security-hardening

The problem is that I rely on `PKCS11Provider /usr/local/lib/opensc-pkcs11.so` in SSH config, I don't ssh-add the key manually. If I try to `ssh-add -s /usr/local/lib/opensc-pkcs11.so`, I get this:

```
Enter passphrase for PKCS#11:
Could not add card "/usr/local/lib/opensc-pkcs11.so": agent refused operation
```

I have no idea what that passphrase is, and no matter what I enter I get the "agent refused operation". The passphrase is not my Yubikey pin or Keychain password or anything else that I can think of.

**Comment by dengert** [comment]
Does anyone who uses OpenSSH and/or uses p11-kit have any insite? @Jakuje 

@danepowell  also note:
```
Slot 1 (0x4): Yubico YubiKey OTP+FIDO+CCID 01
  token label        : PIV_II
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized, **user PIN locked**
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : 00000000
  pin min/max        : 4/8
```
the user PIN is locked. You need to get this unlocked for sure. 
Yubikey tools know how to do this. Who issued you the token?

**Comment by danepowell** [comment]
Yeah the secondary key got locked because of exactly this issue; I was trying to unlock the primary key but OpenSC was taking PIN input for the secondary key and I didn't realize it until it locked out. Anyway, I don't even use PIV on that secondary key so it hasn't been a concern.

**Comment by dengert** [comment]
This might help:
http://lists.mindrot.org/pipermail/openssh-unix-dev/2020-June/038580.html
http://lists.mindrot.org/pipermail/openssh-unix-dev/2020-June/038581.html
It talks about a patch. Does not look like it is in OpenSSH. (It may be in RedHat.)

It expects p11-kit is setup as the OpenSSH pkcs11 module. as per the RedHat  instructions.

On my test  Ubuntu VM with 2 Yubikeys and a real PIV card and SoftHSM 

```
$ p11tool --list-token-urls
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=d289a4f997d27222;token=dee-2
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=b77465002b1885f4;token=DEE-1
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=75b0ec1e3b974305;token=mytoken
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=ffa43bb8db818e0c;token=DEE-1
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=017c072ff770b055;token=dee-3
pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=System%20Trust
pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;serial=818cf26b716ae666;token=Doug_Y8
pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;serial=a84187a97da31764;token=deengert
pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;serial=de0a61813810d7e1;token=ID-One%20PIV%20TEST%20...%20%28Global%20PIN%29
```
p11tool looks at all the tokens p11-kit can see and lists out the URI which would be needed to find each of the tokens. Note the use of `serial=` and `label=`  With p11-kit, the `token=` appears to be the `CN=`  from the first certificate's subjectName. 

Best I can tell, OpenSSH without the patch does not support `serial=` or `label=` But it is worth a try. 

The OpenSSL PKCS11 provider mentioned in the e-mail is under development at https://github.com/latchset/pkcs11-provider

I don't use OpenSSH much these days.

**Comment by frankmorgner** [comment]
We don't have the filter mechanism that you are requesting. Typically, the hosting application (openssh) checks the token for the requested key before accessing the token. OpenSSH [seems to be looking for a specific keyid](https://github.com/openssh/openssh-portable/blob/master/ssh-pkcs11.c#L382). If I'm reading the code correctly, then the keyid is extracted from the (trusted) certificate's key identifier.

**Comment by frankmorgner** [comment]
Just to be clear: Your best option would be to contact the OpenSSH support on how to access the correct key/token.

**Comment by dengert** [comment]
> OpenSSH [seems to be looking for a specific keyid](https://github.com/openssh/openssh-portable/blob/master/ssh-pkcs11.c#L382). If I'm reading the code correctly, then the keyid is extracted from the (trusted) certificate's key identifier.

I agree. But then it looks at the first slot. OpenSC PIV pkcs15/pkcs11 maps the  "X.509 Certificate for PIV Authentication" (NIST sp800-73-*) to  CKA_ID=01 for ever token. So OpenSSH will find it on the first slot. The pubkey, privkey and  certificates are number sequentially in pkcs15-piv.c This has been there since 2004 without problems, and actually made PAM login easier as PAM could always looked for CKA_ID=01. 

[pkcs11-base v3.0](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.pdf) "The CKA_ID field is intended to distinguish among multiple keys. ... The key identifier should also be the same as for the corresponding certificate, if one exists. Cryptoki does not enforce these associations, however." 

https://www.rfc-editor.org/rfc/rfc7512 defines URI scheme for PKCS11.

OpenSSH only implements parts of this, but does not include the "serial=" or 'token=" which could be used to further select the token to be used. 

The p11tool example above shows what the token part of the URI could be used to only look at one token. 

Full implementation of rfc7512 are in p11-kit, and OpenSSL via libp11 and the pkcs11-privider under development. The RedHat documentation appears to indicate that they may have mods to OpenSSH to force the URI strings to be passed to OpenSSL or to p11-kit which would allow the use of "serial=" or 'token=" to get around the issue. 

OpenSSH source defines "USE_OPENSSL_ENGINE" which might also work to have OpenSSL call the libp11 engine, passing the URI provided. 

Note the URI is not passed directly to a PKCS11 module as a URI. The caller needs to parse the URI and  use it to select a token, and then the objects on the token.

**Comment by dengert** [comment]
I disagree for a number of reasons: 

- [RFC7512](https://www.rfc-editor.org/rfc/rfc7512) "This memo specifies a PKCS#11 Uniform Resource Identifier (URI) Scheme for identifying PKCS#11 objects stored in PKCS#11 tokens and also for identifying PKCS#11 tokens, slots, or libraries." This allows for **selecting a token based on the token serial or token label, in addition to the CKA_ID.**    The problem is OpenSSH does not support rfc7512 directly. It can by using an OpenSSL engine such as libp11 or an OpenSSL 3.0 provider such as  https://github.com/latchset/pkcs11-provider which do support rfc7512.

- The PKCS#11 quote above says: "in such a certificate extension"  This would appear to be the the "subjectKeyIdentifier" https://security.stackexchange.com/questions/200295/the-difference-between-subject-key-identifier-and-sha1fingerprint-in-x509-certif Generation of certificates are out of control of OpenSC so certificates may not have a  subjectKeyIdentifier

- Not all smartcards have certificates, especially when the smartcard is used with OpenSSH. 

- Many users may depend on the CKA_ID being unique to the token and have URIs in scripts. This is very handy for login to allow the correct key to be select for a specific token.  

- Changing how OpenSC defines CKA_ID from being unique to the token to something else, will cause people to have to rewrite their URIs. 

It looks like at least PIV and OpenPGP assign simple CKA_IDs starting at 1.

**Comment by dengert** [comment]
@danepowell  Are you willing to try and use [libp11](https://github.com/OpenSC/libp11) which supports  rfc7512?

When you build OpenSSH in `/usr/local`   add to the `configure --with-ssl-engine`
Then make sure libp11 is setup an engine for openssl

You did not say what OS you are using. It most likely has a libp11 package.

Then use ssh-add -e pkcs11 This is where it gets messy, and I have not tried it myself.

**Comment by danepowell** [comment]
I've been tinkering with this and the problem seems to be a little more nuanced. 

If I explicitly add the pkcs11 key to the agent (`ssh-add -s /usr/local/lib/opensc-pkcs11.so`), it works fine even if PIV_II (the lock and undesirable token) is in slot 0. It still somehow knows to unlock and add the Foo PIV token in slot 1.

If I don't explicitly run ssh-add but instead use PKCS11Provider in my ssh config and rely on this to invoke OpenSC when initiating an SSH connection, it first tries the PIV_II token in slot 0, then falls back to the Foo PIV token in slot 1, which then works. I'm afraid I got some wires crossed in tinkering here, as the behavior now is different from before (originally it would outright fail after trying slot 0). Possibly this is because I installed openssh from Brew and copied the OpenSC files into `/usr/local/lib` instead of relying on the symlinks.

Now that this works fine if I explicitly run ssh-add, and it works okay-ish if I use PKCS11Provider in my ssh config, I'm mostly satisfied. I'd still prefer a way to direct OpenSC to the correct token when using PKCS11Provider in my ssh config, but it's not worth the time it would take to debug this further.

**Comment by Jakuje** [comment]
This issue has several levels where it can be solved:

 * First of all, I think yubikeys can be configured to disable different endpoints so if you are not using the CCID endpoint on one yubikey, it should be possible to disable, for example using [Yubikey Manager](https://developers.yubico.com/yubikey-manager-qt/) -- Under "Interfaces" tab you can disable the CCID (Smart Card) USB interface and the second token should not show up for you at all.
 * If the above does not work, for whatever reason, then applications should be able to handle this with PKCS#11 URIs, as mentioned previously. It solves several issues like this, but upstream was ignoring this for years (regardless they have working patches for 5 years we are using in Fedora and RHEL): https://bugzilla.mindrot.org/show_bug.cgi?id=2817
 * One more cumbersome solution might be using p11-kit remoting (even locally), which can allow you filtering keys/slots/objects as a side effect: https://p11-glue.github.io/p11-glue/p11-kit/manual/remoting.html -- you can start p11-kit-server with URI selecting only the key you want, export environment variables and then add p11-kit-client to ssh, which should see only the respective slot/objects. This is cumbersome and sounds fragile, but it might be possible to script/automate it if needed.

**Comment by dengert** [comment]
@Jakuje   I like your patch.  https://bugzilla.mindrot.org/show_bug.cgi?id=2817#c5 addresses a different but important issue.  https://bugzilla.mindrot.org/show_bug.cgi?id=2817#c6
says it works on MacOS.

Could your patch also allow the use of "serial"? 

In @danepowell case two organizations have given him Yubikey tokens and both look like they are using the PIV applet. So turning off CCID would not work. 

The OpenSSH use of OpenSSL's engine or OpenSSL 3.0 provider might also work, as it does not require a patch, but OpenSSH must be configured with `--with-ssl-engine`. Not clear if other distros  configure with  `--with-ssl-engine`.

**Comment by mouse07410** [comment]
@dengert OpenSSL-3+ is moving away from engine API to provider API. More so, the current `libp11` master  doesn't compile on my Macs anymore - I'm not making waves because it's deprecated. 

Would your otherwise-reasonable proposal to use `--with-ssl-engine` work with providers as well?

**Comment by Jakuje** [comment]
It was not clear to me if the "browser-based auth" means PKCS#11 or FIDO2/U2F or something else.
Adding a possibility to filter by serial number to openssh pkcs11 patch should be trivial, please comment on the upstream bug to give it some attention.

The current version can be pulled from Fedora:

https://src.fedoraproject.org/rpms/openssh/blob/rawhide/f/openssh-8.0p1-pkcs11-uri.patch

I will provide a patch for the serial (but serials with yubikeys are also tricky as they require also some setup.

I do not think OpenSSH would work with ssl-engine options -- this was an option to use to accelerate/offload crypto to some openssl-engine, but for using keys, you still need a file or an external pkcs11 module (not through an engine).

**Comment by Jakuje** [comment]
Change adding a support for the serial field is available in here: https://src.fedoraproject.org/rpms/openssh/pull-request/46

**Comment by dengert** [comment]
@Jakuje I have taken the above references to your "pkcs11-URI" and applied then  to Openssh-portable See:  https://github.com/dengert/openssh-portable  branch pkcs11-URI

Using 2 Yubikeys:
```
pkcs11-tool --list-slots
Available slots:
Slot 0 (0x0): Yubico YubiKey OTP+FIDO+CCID 00 00
  token label        : deengert
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : a84187a97da31764
  pin min/max        : 4/8
Slot 1 (0x4): Yubico YubiKey OTP+FIDO+CCID 01 00
  token label        : Doug_Y8
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : 818cf26b716ae666
  pin min/max        : 4/8
```
and stitching slots and testing with Ubuntu sshd on port 2222 using the opensc-pkcs11.so module.
```
sudo /usr/sbin/sshd -p 2222 -d -d -d

./ssh -i "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;token=Doug_Y8;id=01" -v -v -v -p 2222 localhost
./ssh -i "pkcs11:manufacturer=piv_II;serial=818cf26b716ae666;id=01" -v -v -v -l doug -p 2222 localhost

```
It always finds the correct token, and prompts for `Enter PIN for 'Doug_Y8': `

@danepowell you could build OpenSSH from clone of  https://github.com/dengert/openssh-portable and see if this would fix your problem.

**Comment by dengert** [comment]
I have pushed additional changes to https://github.com/dengert/openssh-portable branch pkcs11-URI that also gets the `regress/unittests/pkcs11/test_pkcs11` working:
test_pkcs11: ........................................................ 56 tests ok' on local machine had to use `make -k tests` to get this.  I think there is some problem in unrelated codein openssh tests. 

@Jakuje are you willing to have a look at this? Would this be more acceptable to OpenSSH that the 5 year old patches?

**Comment by Jakuje** [comment]
Sorry for a delay. When implementing the serial patch, I missed that the the serial number is stripped the last character of serial numbers. This was in the meantime fixed in OpenSSH openssh/openssh-portable#406 and the Fedora patch to support pkcs11 uris should work.

Is there something we can or want to do with this issue from the OpenSC point of view? I do not think so so I think we can close this issue.
